File: frontend/src/services/aiScope.ts
import policy from '../../../shared/aiScopePolicy.json';
import type { ClarificationQuestion } from '../types/ai';

export type AiScopeCode = keyof typeof policy.messages;

export class AiScopeError extends Error {
  readonly code: AiScopeCode;

  constructor(code: AiScopeCode) {
    super(policy.messages[code]);
    this.name = 'AiScopeError';
    this.code = code;
  }
}

const injectionPatterns = policy.injectionPatterns.map(pattern => new RegExp(pattern, 'iu'));
const offTopicPatterns = policy.offTopicPatterns.map(pattern => new RegExp(pattern, 'iu'));
const topicPatterns = policy.topicPatterns.map(pattern => new RegExp(pattern, 'iu'));

function normalize(text: string, preserveClauses = false): string {
  const normalized = text.normalize('NFKC').toLowerCase()
    .replace(/[\u200b-\u200f\u202a-\u202e\u2060-\u206f\ufeff]/g, '');
  return (preserveClauses ? normalized.replace(/[\r\n]+/g, '; ') : normalized)
    .replace(/\s+/g, ' ').trim();
}

/** Short clarification answers inherit scope from the draft, never from industry. */
export function assertTaskScope(draft: string, industry = '', answers: string[] = []): void {
  const input = [draft, industry, ...answers];
  const texts = input.map(text => normalize(text));
  if (texts.some(text => injectionPatterns.some(pattern => pattern.test(text)))) {
    throw new AiScopeError('PROMPT_INJECTION');
  }
  if (input.some(text => offTopicPatterns.some(pattern => pattern.test(normalize(text, true)))) ||
      !topicPatterns.some(pattern => pattern.test(texts[0]))) {
    throw new AiScopeError('OFF_TOPIC');
  }
}

/** AI selects fields; the UI receives only reviewed clarification text and IDs. */
export function getClarificationQuestion(field: string): ClarificationQuestion {
  if (!Object.hasOwn(policy.questions, field)) throw new AiScopeError('OFF_TOPIC');
  return {
    id: `q-${field}`,
    field,
    question: policy.questions[field as keyof typeof policy.questions],
  };
}

export function sanitizeQuestions(questions: ClarificationQuestion[]): ClarificationQuestion[] {
  return questions.map(({ field }) => getClarificationQuestion(field));
}

File: backend/services/ai_scope.py
"""Deterministic input rails shared with the browser; no model or network required."""

import json
import re
import unicodedata
from pathlib import Path
from typing import Iterable, Literal


POLICY_PATH = Path(__file__).resolve().parents[2] / "shared" / "aiScopePolicy.json"
POLICY = json.loads(POLICY_PATH.read_text(encoding="utf-8"))
CANONICAL_QUESTIONS = POLICY["questions"]
INJECTION_PATTERNS = tuple(re.compile(pattern, re.IGNORECASE) for pattern in POLICY["injectionPatterns"])
OFF_TOPIC_PATTERNS = tuple(re.compile(pattern, re.IGNORECASE) for pattern in POLICY["offTopicPatterns"])
TOPIC_PATTERNS = tuple(re.compile(pattern, re.IGNORECASE) for pattern in POLICY["topicPatterns"])
WORK_CONTEXT_PATTERNS = tuple(re.compile(pattern, re.IGNORECASE) for pattern in POLICY["workContextPatterns"])
WORK_INTENT_PATTERNS = tuple(re.compile(pattern, re.IGNORECASE) for pattern in POLICY["workIntentPatterns"])
INVISIBLE_CHARACTERS = re.compile(r"[\u200b-\u200f\u202a-\u202e\u2060-\u206f\ufeff]")


class ScopeViolation(Exception):
    def __init__(self, code: Literal["OFF_TOPIC", "PROMPT_INJECTION"]):
        self.code = code
        self.message = POLICY["messages"][code]
        super().__init__(self.message)


def normalize_scope_text(text: str, preserve_clauses: bool = False) -> str:
    """Normalize only for inspection; retain the original text in extracted fields."""
    normalized = unicodedata.normalize("NFKC", text).lower()
    normalized = INVISIBLE_CHARACTERS.sub("", normalized)
    # Injection commands may span lines; only topical rules need clause boundaries.
    if preserve_clauses:
        normalized = re.sub(r"[\r\n]+", "; ", normalized)
    return " ".join(normalized.split())


def ensure_ai_scope(draft: str, industry: str = "", answers: Iterable[str] = ()) -> None:
    """Reject overrides and unrelated requests before checking the draft's domain."""
    texts = (draft, industry, *answers)
    normalized = [normalize_scope_text(text) for text in texts]
    # Check every text for overrides first, regardless of where off-topic text occurs.
    if any(pattern.search(text) for text in normalized for pattern in INJECTION_PATTERNS):
        raise ScopeViolation("PROMPT_INJECTION")
    clauses = [normalize_scope_text(text, preserve_clauses=True) for text in texts]
    if any(pattern.search(text) for text in clauses for pattern in OFF_TOPIC_PATTERNS):
        raise ScopeViolation("OFF_TOPIC")
    # Industry and answer keywords cannot turn an unrelated draft into a business task.
    # Accept everyday work descriptions without demanding software/business jargon.
    has_task_context = any(pattern.search(normalized[0]) for pattern in TOPIC_PATTERNS) or (
        any(pattern.search(normalized[0]) for pattern in WORK_CONTEXT_PATTERNS)
        and any(pattern.search(normalized[0]) for pattern in WORK_INTENT_PATTERNS)
    )
    if not has_task_context:
        raise ScopeViolation("OFF_TOPIC")
